src/transactions/typeTag/typeTagParser.ts

```diff
@@ -0,0 +1,246 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable class-methods-use-this */
+/* eslint-disable max-classes-per-file */
+import {
+  StructTag,
+  TypeTag,
+  TypeTagAddress,
+  TypeTagBool,
+  TypeTagParserError,
+  TypeTagStruct,
+  TypeTagU128,
+  TypeTagU16,
+  TypeTagU256,
+  TypeTagU32,
+  TypeTagU64,
+  TypeTagU8,
+  TypeTagVector,
+} from "./typeTag";
+import { AccountAddress } from "../../core";
+import { Identifier } from "../instances";
+
+function isValidIdentifier(str: string) {
+  return !!str.match(/^[_a-zA-Z0-9]+$/);
+}
+
+function isValidWhitespaceCharacter(char: string) {
+  return !!char.match(/\s/);
+}
+
+function consumeWhitespace(tagStr: string, pos: number) {
+  let i = pos;
+  for (; i < tagStr.length; i += 1) {
+    const innerChar = tagStr[i];
+
+    if (!isValidWhitespaceCharacter(innerChar)) {
+      // If it's not colons, and it's an invalid character, we will stop here
+      break;
+    }
+  }
+  return i;
+}
+
+/**
+ * All types are made of a few parts they're either:
+ * 1. A simple type e.g. u8
+ * 2. A standalone struct e.g. 0x1::account::Account
+ * 3. A nested struct e.g. 0x1::coin::Coin<0x1234::coin::MyCoin>
+ *
+ * There are a few more special cases that need to be handled, however.
+ * 1. Multiple generics e.g 0x1::pair::Pair<u8, u16>
+ * 2. Spacing in the generics e.g. 0x1::pair::Pair< u8 , u16>
```

Comment:
Could you normalize all inputs by removing all whitespace and then re-adding it only after commas?

Reply:
Yeah, also normalization by removing spaces and then reparsing is dangerous, if you jam two words together when you forget to have commas.  (not like anyone should be naming their types like that)